rmw_fastrtps_cpp/src/subscription.cpp

```diff
@@ -128,47 +148,50 @@ create_subscription(
   subscriberParam.topic.topicName = _create_topic_name(qos_policies, ros_topic_prefix, topic_name);
 
   if (!get_datareader_qos(*qos_policies, subscriberParam)) {
-    RMW_SET_ERROR_MSG("failed to get datareader qos");
-    goto fail;
+    return nullptr;
   }
-  info->listener_ = nullptr;
+
   if (create_subscription_listener) {
     info->listener_ = new (std::nothrow) SubListener(info);
     if (!info->listener_) {
       RMW_SET_ERROR_MSG("create_subscriber() could not create subscriber listener");
-      goto fail;
+      return nullptr;
     }
   }
+
   info->subscriber_ = Domain::createSubscriber(participant, subscriberParam, info->listener_);
   if (!info->subscriber_) {
     RMW_SET_ERROR_MSG("create_subscriber() could not create subscriber");
-    goto fail;
+    return nullptr;
   }
   info->subscription_gid_ = rmw_fastrtps_shared_cpp::create_rmw_gid(
     eprosima_fastrtps_identifier, info->subscriber_->getGuid());
-  rmw_subscription = rmw_subscription_allocate();
+
+  rmw_subscription_t * rmw_subscription = rmw_subscription_allocate();
   if (!rmw_subscription) {
     RMW_SET_ERROR_MSG("failed to allocate subscription");
-    goto fail;
+    return nullptr;
   }
+  auto cleanup_subscription = rcpputils::make_scope_exit(
+    [rmw_subscription]() {
+      rmw_free(const_cast<char *>(rmw_subscription->topic_name));
+      rmw_subscription_free(rmw_subscription);
+    });
+
   rmw_subscription->implementation_identifier = eprosima_fastrtps_identifier;
   rmw_subscription->data = info;
-  rmw_subscription->topic_name = rcutils_strdup(topic_name, rcutils_get_default_allocator());
```

Comment:
Why are you replacing the user of `rcutils_strdup`? It seems better than manually allocating and memcpy'ing.

Reply:
To at least be consistent with the rest of this implementation and others. We still owe ourselves a discussion on how allocations should be made.

Reply:
Unless this unifies how we do string duplication everywhere in the package, then I think this is a step backwards and kind of unrelated to this pull request right? I just don't see why we're doing this in this pull request. The previous state wasn't wrong was it?

Reply:
It does for API that does not explicitly take an `rcutils_allocator_t` argument (see https://github.com/ros2/rmw_fastrtps/search?q=rcutils_strdup&unscoped_q=rcutils_strdup). 

But provided that this isn't part of the API spec, you're right in that it doesn't have to be in this PR. I'll remove it.

Reply:
See https://github.com/ros2/rmw_fastrtps/pull/419/commits/791a3308eb4aa92fc90d9b24ef60a24b636ce210